tests/test_runner.py

```diff
@@ -0,0 +1,166 @@
+        myrunner = runner.BeakerRunner(**DEFAULT_ARGS)
+        result = myrunner.getxml({'TEST': 'replaced text'})
+        expected_xml = self.test_xml.replace("##TEST##", "replaced text")
+        self.assertEqual(result, expected_xml)
+
+    @mock.patch('subprocess.Popen')
+    def test_getresultstree(self, mock_popen):
+        """Ensure getresultstree() works"""
+        test_xml = "<xml><test>TEST</test></xml>"
+        mock_popen.return_value.returncode = 0
+        mock_popen.return_value.communicate.return_value = (test_xml, '')
+        myrunner = runner.BeakerRunner(**DEFAULT_ARGS)
+        result = myrunner.getresultstree(jobid=0)
+        self.assertEqual(next(x.text for x in result.iter('test')), 'TEST')
+
+    @mock.patch('subprocess.Popen')
+    def test_dumpjunitresults(self, mock_popen):
+        """Ensure dumpjunitresults() works"""
+        mock_popen.return_value.returncode = 0
+        mock_popen.return_value.communicate.return_value = (
+            'stdout',
+            'stderr'
+        )
+
+        junit_dir = tempfile.mkdtemp()
+        myrunner = runner.BeakerRunner(**DEFAULT_ARGS)
+        myrunner.dumpjunitresults("J:00001", junit_dir)
+        expected_file = "{}/j_00001.xml".format(junit_dir)
+        self.assertTrue(os.path.exists(expected_file))
+
+    @mock.patch('skt.runner.BeakerRunner.getresultstree')
+    def test_getconsolelog(self, mock_getresultstree):
+        """Ensure getconsolelog() works"""
+        # Mock up a beaker XML reply
+        beaker_xml = '{}/assets/beaker_results.xml'.format(SCRIPT_PATH)
+        with open(beaker_xml, 'r') as fileh:
+            mocked_xml = fileh.read()
+        mock_getresultstree.return_value = etree.fromstring(mocked_xml)
+
+        myrunner = runner.BeakerRunner(**DEFAULT_ARGS)
+        result = myrunner.getconsolelog()
+        self.assertEqual(result, "http://example.com/")
+
+    def test_forget_cid_withj(self):
+        """Ensure _forget_cid() works with J type jobs"""
+        # pylint: disable=protected-access
+        myrunner = runner.BeakerRunner(**DEFAULT_ARGS)
+        myrunner.jobs = ["J:00001"]
+        myrunner.j2r = {"J:00001": ["R:00001"]}
+        myrunner.recipes = ["R:00001"]
+        result = myrunner._forget_cid("J:00001")
+        self.assertIsNone(result)
+        self.assertEqual(myrunner.jobs, [])
+        self.assertEqual(myrunner.recipes, [])
+
+    def test_forget_cid_withr(self):
+        """Ensure _forget_cid() works with R type jobs"""
+        # pylint: disable=protected-access
+        myrunner = runner.BeakerRunner(**DEFAULT_ARGS)
+        myrunner.jobs = ["J:00001"]
+        myrunner.j2r = {"J:00001": ["R:00001"]}
+        myrunner.recipes = ["R:00001"]
+        result = myrunner._forget_cid("R:00001")
+        self.assertIsNone(result)
+        self.assertEqual(myrunner.recipes, [])
+
+    def test_forget_cid_bad_job(self):
+        """Ensure _forget_cid() fails with unknown jobs"""
```

Comment:
I expect this should be "fails with invalid taskspec".

Reply:
Done